src/Sulu/Component/HttpCache/EventSubscriber/InvalidationSubscriber.php

```diff
@@ -57,25 +93,125 @@ public function __construct(
+        }
+    }
+
+    /**
+     * Invalidates all urls in the locale of the document of when a document gets unpublished.
+     * This method is executed before the actual unpublishing of the document because the document must still
+     * be published to gather the urls of the document.
+     *
+     * @param UnpublishEvent $event
+     */
+    public function invalidateDocumentBeforeUnpublishing(UnpublishEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof ResourceSegmentBehavior || !$document->getPublished()) {
+            // no need to invalidate the path to the document if it is not published currently
+            return;
+        }
+
+        // invalidate urls of the locale of the document
+        $locale = $this->documentInspector->getLocale($document);
+        foreach ($this->getLocaleUrls($document, $locale) as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Invalidates all urls in all locales of the document when a document gets removed.
+     * This method is executed before the actual removing of the document because the document must still
+     * exist to gather the urls of the document.
+     *
+     * @param RemoveEvent $event
+     */
+    public function invalidateDocumentBeforeRemoving(RemoveEvent $event)
     {
         $document = $event->getDocument();
 
-        if (!$document instanceof StructureBehavior) {
+        if (!$document instanceof ResourceSegmentBehavior) {
             return;
         }
 
-        $structureBridge = $this->structureManager->wrapStructure(
-            $this->documentInspector->getMetadata($document)->getAlias(),
-            $this->documentInspector->getStructureMetadata($document)
-        );
+        // invalidate urls of all locales of the document
+        $urls = [];
+        foreach ($this->documentInspector->getPublishedLocales($document) as $locale) {
+            $urls = array_merge($urls, $this->getLocaleUrls($document, $locale));
+        }
 
-        $structureBridge->setDocument($document);
+        foreach ($urls as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Returns all urls of the given locale which are associated with the given document.
+     * The returned array contains all current urls and all history urls.
+     * The returned urls can contain placeholders (eg {host}).
+     *
+     * @param ResourceSegmentBehavior $document
+     * @param $locale
+     *
+     * @return array Urls of the given locale which are associated with the given document
+     */
+    private function getLocaleUrls(ResourceSegmentBehavior $document, $locale)
+    {
+        $documentUuid = ($document instanceof UuidBehavior) ? $document->getUuid() : null;
+        $webspace = ($document instanceof WebspaceBehavior) ? $document->getWebspaceName() : null;
+
+        if (!$document || !$locale || !$documentUuid) {
```

Comment:
`!$document` is not necessary, because the method signature already doesn't allow null.